compilers/concrete-compiler/compiler/lib/Dialect/TFHE/Transforms/TFHECircuitSolutionParametrization.cpp

```diff
@@ -203,15 +204,152 @@ class TFHEParametrizationTypeResolver : public TypeResolver {
       std::optional<CircuitSolutionWrapper> solution)
       : solution(solution) {}
 
+  // Adds constraints that ensure that the argument types of `funcOp`
+  // and the operand types of `callOp` are equal and that the result
+  // types of all return operations of `funcOp` are equal to the
+  // result types of `callOp`
+  void addFuncAndCallConstraints(TypeConstraintSet<> &cs,
+                                 mlir::func::FuncOp funcOp,
+                                 mlir::func::CallOp callOp) {
+    // Ensure that the operands of the call op have the same
+    // type as the function arguments
+    for (auto [callOperand, funcArgument] :
+         llvm::zip_equal(callOp->getOperands(), funcOp.getArguments())) {
+      cs.addConstraint<DynamicSameTypeConstraint<DynamicFunctorYield>>(
+          [callOperand = callOperand]() { return callOperand; },
```

Comment:
Agreed that this is less readable. Unfortunately, this is required, since the C++ standard prohibits captures of expressions from structured bindings explicitly.